namerd/storage/etcd/src/main/scala/io/buoyant/namerd/storage/etcd/EtcdDtabStore.scala

```diff
@@ -0,0 +1,121 @@
+              }
+
+            case Action.CompareAndDelete | Action.Delete | Action.Expire =>
+              state match {
+                case Activity.Ok(namespaces) =>
+                  state = Activity.Ok(namespaces - namespace(nodeOp.node.key))
+                case _ =>
+              }
+
+            case action =>
+          }
+          updates.update(state)
+        case Throw(e) =>
+          updates.update(Activity.Failed(e))
+      }
+    }
+    Activity(run)
+  }
+
+  def create(ns: Ns, dtab: Dtab): Future[Unit] = {
+    val buf = Buf.Utf8(dtab.show)
+    root.key(Path.Utf8(ns)).create(Some(buf)).rescue {
+      case ApiError(ApiError.NodeExist, _, _, _) =>
+        Future.exception(new DtabNamespaceAlreadyExistsException(ns))
+    }.unit
+  }
+
+  def delete(ns: Ns): Future[Unit] = {
+    root.key(Path.Utf8(ns)).delete().rescue {
+      case ApiError(ApiError.KeyNotFound, _, _, _) =>
+        Future.exception(new DtabNamespaceDoesNotExistException(ns))
+    }.unit
+  }
+
+  def update(ns: Ns, dtab: Dtab, version: Version): Future[Unit] = {
+    val Buf.Utf8(vstr) = version
+    val buf = Buf.Utf8(dtab.show)
+    val key = root.key(Path.Utf8(ns))
+    Future(vstr.toLong).rescue {
+      case e: NumberFormatException =>
+        Future.exception(new DtabVersionMismatchException)
+    }.flatMap { index =>
+      key.compareAndSwap(buf, prevIndex = Some(index))
+    }.rescue {
+      case ApiError(ApiError.TestFailed, _, _, _) =>
+        Future.exception(new DtabVersionMismatchException)
+      case ApiError(ApiError.KeyNotFound, _, _, _) =>
+        Future.exception(new DtabNamespaceDoesNotExistException(ns))
+    }.unit
+  }
+
+  def put(ns: Ns, dtab: Dtab): Future[Unit] =
+    root.key(Path.Utf8(ns)).set(Some(Buf.Utf8(dtab.show))).unit
+
+  def observe(ns: Ns): Activity[Option[VersionedDtab]] = {
+    val run = Var.async[Activity.State[Option[VersionedDtab]]](Activity.Pending) { updates =>
+
+      val key = root.key(Path.Utf8(ns))
+      key.events().respond {
+        case Return(nodeOp) => nodeOp.action match {
+          case Action.CompareAndDelete | Action.Delete | Action.Expire =>
+            updates.update(Activity.Ok(None))
+          case Action.CompareAndSwap | Action.Create | Action.Get | Action.Set | Action.Update =>
+            nodeOp.node match {
+              case Node.Data(_, _, _, _, Buf.Utf8(dtabStr)) =>
+                val version = Buf.Utf8(nodeOp.node.modifiedIndex.toString)
+                val dtab = Dtab.read(dtabStr)
+                updates.update(Activity.Ok(Some(VersionedDtab(dtab, version))))
+              case dir: Node.Dir =>
+                updates.update(Activity.Failed(new IllegalStateException(s"${key.path.show} is not a data node")))
+
+            }
+          case _ =>
+        }
+
+        case Throw(ApiError(ApiError.KeyNotFound, _, _, _)) =>
+          updates.update(Activity.Ok(None))
+        case Throw(e) =>
+          updates.update(Activity.Failed(e))
```

Comment:
fwiw, the `states() = Activity...` form is fairly idiomatic throughout the rest of our codebase, as well as finagle's...  i don't feel strongly, though